libraries/ml/src/spe.ts

```diff
@@ -176,3 +176,89 @@ export class PSPEBase extends SPEBase {
     return coordinates;
   }
 }
+
+/**
+ * Implements modified stochastic proximity embedding.
+ *
+ * @export
+ * @class SPE
+ * @link doi:10.1002/jcc.10234
+ */
+export class OriginalSPE extends SPEBase {
+  /**
+   * Embeds the vectors given into a two-dimensional space using a modified update rule.
+   *
+   * @param {Vectors} vectors D-dimensional coordinates.
+   * @return {Coordinates} SPE coordinates in D space.
+   */
+  protected radiusPercent: number;
+  protected maxDistance: number;
+  protected maxDistanceSteps: number;
+
+  constructor(options?: Options) {
+    super(options);
+    this.cycles = options?.cycles ?? 1e3;
+    this.steps = options?.steps ?? 100000;
+    this.radiusPercent = options?.radiusPercent ?? 1.0;
+    this.maxDistance = options?.maxDistance ?? null;
+    this.maxDistanceSteps = options?.maxDistanceSteps ?? null;
+  }
+
+  public embed(vectors: Vectors): Coordinates {
+    const nItems = vectors.length;
+    const areaWidth = 40;
+    // Initialize the D-dimensional coordinates of the N points.
+    const coordinates = fillRandomMatrix(nItems, OriginalSPE.dimension, areaWidth);
+
+    this.initDistance(vectors);
+
+    if (this.maxDistanceSteps == null) {
+      this.maxDistanceSteps = nItems * Math.floor((nItems - 1) / 2);
+    }
+    if (this.maxDistance == null) {
+      this.maxDistance = -1e37;
```

Comment:
1. What is the constant for? 
2. And why is it equals that number? 
3. Why is it declared within `this` scope?

Reply:
I believe `maxDistanceSteps` parameter is needed to find the most optimal `maxDistance` parameter. I am not the author of this algorithm. I just referred to the algorithm from the core. I can send you the original code, if you want. Also, `maxDistanceSteps` in `this` scope, because it could be set in the constructor.